Optimize default encoder

837951602 opened the pull request:
with String.fromCharCode
Not sure if it's a temporary fix, so no recompile

fasttime commented:
Thanks for working on this! I'm glad not to be the only one around here. I haven't tested it but there are some issues that need to be fixed:

* `fromCodePoint` works mostly like `fromCharCode` and it's cheaper sometimes. That's what the constant `FROM_CHAR_CODE` is for.
* For char codes 0 - 9 it's better to have a number rather than a string as an argument.
* Depending on the character, `charEncodeByCharCode` can be a better choice than the other default char encoder even without `CAPITAL_HTML`. For example with characters like "\ubbbb". Because of this I think it's probably better to test both alternatives and choose the cheapest one, rather than having it hard-coded for some particular features. This is something that should be done in `defaultResolveCharacter`.

That's all for now, but I'll look back at this later.

837951602 commented:
fromCodePoint may be different but it's not used yet. 
`'\u86a0!\u86a0'` 26003c
`[].filter.constructor('return String.fromCharCode('+'100000false33false100000'.split(false)+')')()` 25949c

[FROM_CODE_POINT, BARPROP]
`'\uD87E\uDC04!'` 20688c
`'[].filter.constructor('return String.fromCodePoint('+'194564false33'.split(false)+')')()'` 20632c

Here I still use the 'String' so they are comparable

837951602 commented:
Is there function that choose a shorter string? I think it should be commomly used

fasttime commented:
Thanks. Since those are two unrelated optimizations here (char encoding by char code and `(isNaN+false).constructor`) it would make it easier to submit two PRs with different commits and titles (don't care about the conflicts).

The way to choose whatever definition resolves shorter is using `findOptimalSolution`, but this is not necessary here. Just add a constant in definitions.js, maybe like this:

Dclare `var OPTIMAL_STRING` a the top of the file just under `OPTIMAL_B`, then define it down below:

```js
OPTIMAL_STRING =
[
    define('(isNaN+false).constructor'),
    define(' String', CHAR_HTML, ENTRIES_OBJ)
];
```

In encoder-ext.js you can do:

```js
var optimalStr = this.findDefinition(OPTIMAL_STRING);

...

output =
    this.resolveConstant('Function') +
    '(' +
    this.replaceString('return' + optimalStr + '.' + fromCharCode + '(', true) +
    '+' +
    charCodeArrayStr +
    ...
```

(Untested, but way to go).

`findDefinition` just looks for matching features from the last to the first, so ` String` is chosen when both `CHAR_HTML, ENTRIES_OBJ` are available, otherwise `(isNaN+false).constructor`.

You also need to add `OPTIMAL_STRING` in the `global` comment at the very top of encoder-ext.js.
Then you can build using `gulp`: this is much faster than `gulp full`, which is only required when single character solutions change.

If you get any errors, try looking at the messages in the console. As I said, I haven't tested this, but that's more or less how it should be done. Of course tell me if you have any problems.

837951602 commented:
Oh, I submitted on 837951602/JScrewIt/master, didn't realize it just go here.
I think OPTIMAL_STRING makes me think of ''.constructor(as code rather than text)
and when seen may wonder why a space is there
so I made a OPTIMAL_RETURN_STRING

837951602 commented:
encode('"(false+isNaN).constructor"',['CAPITAL_HTML', 'ENTRIES_OBJ', 'V8_SRC']).length
1207
encode('" String"',['CAPITAL_HTML', 'ENTRIES_OBJ', 'V8_SRC']).length
1204

So near they are, can "(false+isNaN).constructor" be better than String with both 'CAPITAL_HTML', 'ENTRIES_OBJ'?

fasttime commented:
Thanks! But you should really run `gulp` to build. I'm getting a bunch of validation errors:

```
/Users/noone/GitHub/JScrewIt/src/lib/definitions.js
   218:24  error  Trailing spaces not allowed     no-trailing-spaces
   219:50  error  Trailing spaces not allowed     no-trailing-spaces
   220:18  error  Trailing spaces not allowed     no-trailing-spaces
   225:14  error  Infix operators must be spaced  space-infix-ops
   225:15  error  Trailing spaces not allowed     no-trailing-spaces
  1503:6   error  Missing semicolon               semi

/Users/noone/GitHub/JScrewIt/src/lib/encoder-ext.js
  395:61  error  'OPTIMAL_RETURN_STRING' is not defined  no-undef
```

I think these are all self-explaining. The warning about `OPTIMAL_RETURN_STRING` not being defined is because it's not listed in the `global` comment at the top of encoder-ext.js with the other constants.

Meanwhile, this is the next step to make sure we have optimal definitions for all feature combinations:

* In debug.js add `OPTIMAL_RETURN_STRING` to the global comment at the top and then down below in the section `Exported entries` (similar to `OPTIMAL_B`).
* In build/verify.js add a new verification entry. Again, this is similat to OPTIMAL_B, but with definitions for `OPTIMAL_RETURN_STRING`, and with `replaceString` instead of `resolveChar`.

Now running `node build/verify.js OPTIMAL_RETURN_STRING` should list feature combinations that are not optimal, like 'CAPITAL_HTML', 'ENTRIES_PLAIN', 'IE_SRC', 'FILL'.
When this is done, we'll see what we can do to improve our definitions. Probably it will be enough to add new entries to the `OPTIMAL_RETURN_STRING` array with new features.

837951602 commented:
Gulp full Error:
[19:47:02] Finished 'test' after 34 s
[19:47:02] Starting 'scan-char-defs'...
scanning character definitions [============        ] 60% 75.4s
[19:48:57] 'scan-char-defs' errored after 1.9 min
[19:48:57] SyntaxError: Circular reference detected: "P" < "P"
    at Encoder.callResolver (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.js:4682:27)
    at Encoder.resolveCharacter (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.js:5066:
22)
    at Encoder.encoder.resolveCharacter (C:\TempCodes\JScrewIt-2.6\build\char-ma
p.js:183:45)
    at Encoder.replaceString (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.js:5012:37)

    at replaceString (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.js:4517:39)
    at Encoder.replaceExpressUnit (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.js:486
5:33)
    at Encoder.replaceExpr (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.js:4819:36)
    at Encoder.charEncodeByCharCode (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.js:2
119:21)
    at Encoder.defaultResolveCharacter (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.j
s:4715:55)
    at Encoder.defaultCharDefinition (C:\TempCodes\JScrewIt-2.6\lib\jscrewit.js:
2258:33)
[19:48:57] 'full' errored after 2.57 min

fasttime commented:
Most likely because 'P' has a default definition `defineDefaultChar('P')`, and that's trying to use `fromCodePoint` which needs the char 'P', and so on. Possible fix: remove `defineDefaultChar('P')` and add `unescape("%50"),` as first entry.

837951602 commented on lib/jscrewit.js:
```diff
@@ -5971,7 +5973,6 @@ var wrapWithEval;
                 }
                 else
                 {
-                    var return_String = this.findDefinition(OPTIMAL_RETURN_STRING);
```
Cuz I thought I broke the system, I redownload and made changes, so there may be some resorting differences

837951602 commented:
Changing 'P's defineDefaultChar('P') makes it worse on ARRAY_ITERATOR, ATOB

837951602 commented:
https://github.com/fasttime/JScrewIt/issues/3#issuecomment-287146433 Make it try to directly use comma first, and let the comma optimize in the String part?

fasttime commented:
This looks already good. Just a couple of quirks:

* There is no need to compute `return_String` at the top of the function while it's only used in one branch. This should be moved inside the `else` block where it is used.
* Feature order in definition entries does not matter but I try to stay consistent and sort lexicographically. Hence <strike>`ENTRIES_OBJ, CAPITAL_HTML`</strike> -> `CAPITAL_HTML, ENTRIES_OBJ`.
* Also, I'd suggest you commit up-to-date versions of jscrewit.js, jscrewit.min.js and output.txt every time those file changes. This makes it easier to follow what's going on in a set of commits.

Then there is this thing that `charEncodeByCharCode` solutions are not being tested. This is my bad since I wasn't smart enough when I wrote unit tests to allow for other default character encodings. I'll see how I can improve that.

fasttime commented:
The cost of commas depends on the available features. The current implementation only optimizes with `concat`, but not with `slice`. I hope this can be improved with a new optimizer, but this is a problem we don't need to care about here yet.

fasttime commented:
Great, thanks! This is merged in the update branch now. Just made some minor style changes: https://github.com/fasttime/JScrewIt/commit/60c775bc6faff664aec9f4d44fb096df0a6aa5b4
I'll leave this open until I update the unit tests.

837951602 commented:
That's different when comma is treated as a single character or a spliter.
[single char]true,11,true < 1,11,1
[spliter] true,11,true > 1,1 1,1

fasttime commented:
Sometimes `charEncodeByCharCode` was generating better results then the default unescape encoders even without the feature `CAPITAL_HTML`. So I decided to merge both encoding methods and pick up whatever output is shorter: https://github.com/fasttime/JScrewIt/blob/3455d52ec8c08e9017da6dea95eaa01b98ebc611/src/lib/definitions.js#L205-L219. The downside is that output.txt now takes painfully long to generate. I still have to think of how this can be improved.

The code coverage issues are also patched now. Not the smartest patch though, but it's ok for now.
